cql2pgjson/src/main/java/org/folio/cql2pgjson/CQL2PgJSON.java

```diff
@@ -404,7 +430,74 @@ private SqlSelect toSql(CQLSortNode node) throws QueryValidationException {
       .append(wrapInLowerUnaccent(vals.getIndexText(), modifiers)).append(desc);
     }
-    return new SqlSelect(where, order.toString());
+    if (column == null) {
+      return new SqlSelect(where, order.toString(), null);
+    }
+    String fullQuery = null;
+    if (node.getSortIndexes().size() == 1 && limit > 0 && offset >= 0) {
+      ModifierSet modifierSet = node.getSortIndexes().get(0);
+      String field = modifierSet.getBase();
+
+      String lessGreater = ">";
+      String ascDesc = "ASC";
+      CqlModifiers modifiers = new CqlModifiers(modifierSet);
+      if (modifiers.getCqlSort() == CqlSort.DESCENDING) {
+        lessGreater = "<";
+        ascDesc = "DESC";
+      }
+      String wrappedColumn;
+      String cutWrappedColumn;
+
+      DbIndex dbIndex = dbIndexMap.computeIfAbsent(field, f -> DbSchemaUtils.getDbIndex(dbTable, f));
+      if (dbIndex.isForeignKey() || "id".equals(field)) {
+        cutWrappedColumn = wrappedColumn = field;
+      } else {
+        IndexTextAndJsonValues vals = getIndexTextAndJsonValues(field);
+        wrappedColumn = wrapInLowerUnaccent(vals.getIndexText(), modifiers);
+        cutWrappedColumn = wrapForLength(wrapInLowerUnaccent(vals.getIndexText(), modifiers));
+      }
+
+      int optimizedSqlSize = 1000;
+      fullQuery = " WITH "
+          + " headrecords AS ("
+          + "   SELECT " + column + ", (" + wrappedColumn + ") AS data_column FROM " + table
+          + "   WHERE (" + where + ")"
+          + "     AND " + cutWrappedColumn + lessGreater
+          + "             ( SELECT " + cutWrappedColumn
+          + "               FROM " + table
+          + "               ORDER BY " + cutWrappedColumn + " " + ascDesc
+          + "               OFFSET " + optimizedSqlSize + " LIMIT 1"
+          + "             )"
+          + "   ORDER BY " + cutWrappedColumn + " " + ascDesc
+          + LIMIT_STR + limit + OFFSET_STR + offset
+          + " ), "
+          + " allrecords AS ("
+          + "   SELECT " + column + ", " + wrappedColumn + " AS data_column FROM " + table
+          + "   WHERE (" + where + ")"
+          + "     AND (SELECT COUNT(*) FROM headrecords) < " + limit
+          + " )"
+          + " SELECT " + column + ", data_column"
+          + "   FROM headrecords"
+          + "   WHERE (SELECT COUNT(*) FROM headrecords) >= " + limit
+          + " UNION"
+          + " (SELECT " + column + ", data_column"
+          + "   FROM allrecords"
+          + "   ORDER BY data_column " + ascDesc
+          + LIMIT_STR + limit + OFFSET_STR + offset
+          + " )"
+          + " ORDER BY data_column " + ascDesc;
+    } else {
+      StringBuilder sql = new StringBuilder(SELECT_STR + column + FROM_STR + table + WHERE_STR + where
+          + ORDER_BY + order.toString());
+      if (offset > 0) {
+        sql.append(OFFSET_STR + offset);
+      }
+      if (limit >= 0) {
+        sql.append(LIMIT_STR + limit);
+      }
+      fullQuery = sql.toString();
```

Comment:
```suggestion
      StringBuilder sql = new StringBuilder()
          .append(SELECT_STR).append(column)
          .append(FROM_STR).append(table)
          .append(WHERE_STR).append(where)
          .append(ORDER_BY).append(order.toString());
      if (offset > 0) {
        sql.append(OFFSET_STR).append(offset);
      }
      if (limit >= 0) {
        sql.append(LIMIT_STR).append(limit);
      }
      fullQuery = sql.toString();
```
StringBuilder has memory and performance advantages, but only if you replace `+` String concatenation by ` append()`.

Reply:
This is almost the same code as in toSql(CQLNode). Consider to extract it into a method
```
static String sqlBuilder(CharSequence column, CharSequence table, CharSequence where, CharSequence order, int offset, int limit)
```
or into an SqlSelect constructor
```
public SqlSelect(CharSequence column, CharSequence table, CharSequence where, CharSequence order, int offset, int limit)
```